bitcoin/src/witness_program.rs

```diff
@@ -20,29 +20,54 @@
 //! If the version byte is 0, but the witness program is neither 20 nor 32 bytes, the script must fail.
 //!
 
-use hex;
-use std::fmt;
 use std::str::FromStr;
 
+use wagyu_model::AddressError;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct WitnessProgram {
     /// The version byte
     pub version: u8,
     /// The witness program bytes
     pub program: Vec<u8>,
 }
 
+#[derive(Debug, Fail, PartialEq, Eq)]
+pub enum WitnessProgramError {
+    #[fail(display = "expected program with length at least 2, at most 40, actual program was length {}", _0)]
+    InvalidProgramLength(usize),
+
+    #[fail(display = "invalid program length {} for script version {}", _0, _1)]
+    InvalidProgramLengthForVersion(usize, u8),
+
+    #[fail(display = "expected version no greater than 16")]
+    InvalidVersion(u8),
+
+    #[fail(display = "expected program length {} to equal length {} as specified by input", _0, _1)]
+    MismatchedProgramLength(usize, usize),
+
+    #[fail(display = "error decoding program from hex string")]
+    ProgramDecodingError,
+}
+
+impl From<WitnessProgramError> for AddressError {
+    fn from(error: WitnessProgramError) -> Self {
+        AddressError::Crate("WitnessProgram", format!("{:?}", error))
+    }
+}
+
 impl WitnessProgram {
-    /// Returns a new witness program given a program with a version byte, data size, and data.
-    pub fn new(program: &[u8]) -> Result<WitnessProgram, &'static str> {
+    /// Returns a new witness program given a program with a version byte (where 0 <= version <= 16), data size, and data.
```

Comment:
`with a version byte (where 0 <= version <= 16)`

Please explain your logic.

Reply:
for a witness program, the first byte specifies the witness program version. The version is a value between 0 and 16 inclusive.

Reply:
A byte is 8 bits, (2^8)-1 == 255